### PR TITLE
Ripe cloud mode destination

### PR DESCRIPTION
**Pull request by simpajj**
This is the implementation of the Ripe cloud mode destination, complementing the device-mode one implemented in https://github.com/segmentio/action-destinations/pull/921

- [x] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [x] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [ ] [Segmenters] Tested in the staging environment

### CODE REVIEW

File: packages/browser-destinations/src/destinations/ripe/index.ts [hunk 1]
```diff
@@ -23,8 +23,8 @@ declare global {
 }
 
 export const destination: BrowserDestinationDefinition<Settings, RipeSDK> = {
-  name: 'Ripe',
-  slug: 'actions-ripe',
+  name: 'Ripe Device Mode (Actions)',
+  slug: 'actions-ripe-web',
```
Comment (rhall-twilio) [inline packages/browser-destinations/src/destinations/ripe/index.ts:27]: It's unfortunate but we don't currently support renaming the slug as this is a backwards compatibility breaking change.

**Comment by joe-ayoub-segment** [comment]
Hi @simpajj , 
I hope you are doing well! 

I've been assigned to work with you on this PR. 

I was wondering if you'd be open to meeting with me to discuss the use-case for this Destination, as well as how Ripe works? If I can get an overview of the Ripe product and the intent for your Destination I will be able to provide you with far better feedback for your PR. It's always easier to fix potential issues before code goes out to Production than it is afterwards, not that I'm expecting to find any issues of course :).   

Ideally I would have preferred to have met with you before you started writing code, but this new process is something that we've just started rolling out this week. 

I know you are keen to get your new Destination deployed ASAP, so would it be possible to meet with me some time this week? 

I'm available Thu 5th from 14:00 to 16:00 UK time and on Fri 6th from 10:00 to 18:00 UK time. Would either of these slots work for you? 

Kind regards,
Joe

File: packages/browser-destinations/src/destinations/ripe/index.ts (hunk 1 above)
Comment (simpajj) [reply packages/browser-destinations/src/destinations/ripe/index.ts:27]: Ah, my bad - thanks! I'll amend the commit to just change the name.

**Comment by simpajj** [comment]
> Hi @simpajj , I hope you are doing well!
> 
> I've been assigned to work with you on this PR.
> 
> I was wondering if you'd be open to meeting with me to discuss the use-case for this Destination, as well as how Ripe works? If I can get an overview of the Ripe product and the intent for your Destination I will be able to provide you with far better feedback for your PR. It's always easier to fix potential issues before code goes out to Production than it is afterwards, not that I'm expecting to find any issues of course :).
> 
> Ideally I would have preferred to have met with you before you started writing code, but this new process is something that we've just started rolling out this week.
> 
> I know you are keen to get your new Destination deployed ASAP, so would it be possible to meet with me some time this week?
> 
> I'm available Thu 5th from 14:00 to 16:00 UK time and on Fri 6th from 10:00 to 18:00 UK time. Would either of these slots work for you?
> 
> Kind regards, Joe

Hi @joe-ayoub-segment! Thanks for the reply.

Sure thing! I'm available at 14:00 tomorrow UK time if that still works for you. Looking forward to meeting you! 

/Simon

File: packages/browser-destinations/src/destinations/ripe/index.ts (hunk 1 above)
Comment (simpajj) [reply packages/browser-destinations/src/destinations/ripe/index.ts]: Amended now :)

**Comment by joe-ayoub-segment** [comment]
hi @simpajj , 2pm UK time today is great.

If you send me your email I'll forward you on an invite. 
Either way, here's the URL to dial into. 

Video call link: https://meet.google.com/jxy-vdxy-dnc

**Comment by simpajj** [comment]
> hi @simpajj , 2pm UK time today is great.
> 
> If you send me your email I'll forward you on an invite. Either way, here's the URL to dial into.
> 
> Video call link: https://meet.google.com/jxy-vdxy-dnc

Good stuff! That's ok, I'll add it to my calendar. See you later.

**Comment by joe-ayoub-segment** [comment]
Notes from meeting with Simon earlier today. Items to address: 

- Most fields should have default values
- Add testAuthentication method to validate the API key
- Change defaultSubscription from 'name' to 'type'
- Change timestamp value from now() to Segment root.timestamp
- Support page defaults for path, url etc.
- Suggest handling reserved group traits https://segment.com/docs/connections/spec/group/
- Review and decide if Alias is really needed, and how it should work https://segment.com/docs/connections/spec/alias/

**Comment by simpajj** [comment]
> Notes from meeting with Simon earlier today. Items to address:
> 
>     * Most fields should have default values
> 
>     * Add testAuthentication method to validate the API key
> 
>     * Change defaultSubscription from 'name' to 'type'
> 
>     * Change timestamp value from now() to Segment root.timestamp
> 
>     * Support page defaults for path, url etc.
> 
>     * Suggest handling reserved group traits https://segment.com/docs/connections/spec/group/
> 
>     * Review and decide if Alias is really needed, and how it should work https://segment.com/docs/connections/spec/alias/

Thanks Joe! I've addressed these now, apart from the one about handling reserved group traits, since we save all of the traits and only use them for internal purposes, such as enrichment.

We also briefly discussed the reason for implementing `alias` and it was simply to maintain compatibility with the analytics.js API, but it's not something that we need since we handle aliasing in our backend. I don't mind keeping it, but let me know if you'd rather have me remove it.

**Comment by joe-ayoub-segment** [comment]
Hi Simon, thanks for your time just now. As discussed, final feedback below: 

Page event properties should be optional 
Page name should not be required 
Be specific about which fields are required or not, and which allow nulls 
Update description for anonymousId field. Not anonymized…
Remove group() and alias() handler code. As it doesn’t do anything in Ripe yet. 
Remember to remover the presets for group and alias
Update the testAuthentication method to return a boolean. E.g. 
     const res = await request<RefreshTokenResponse>('https://us.1plusx.io/api/auth/token', {
       method: 'POST',
       headers: {
         'Content-Type': 'application/json',
         Authorization: `Basic ${Buffer.from(settings.key_id + ':' + settings.client_secret).toString('base64')}`
       },
       body: '{"grant_type":"client_credentials"}'
     })
     if (res.status == 200) {
       return { accessToken: res.data.access_token }
     } else {
       throw new Error(res.status + res.statusText)
     }
   },

File: packages/destination-actions/src/destinations/ripe/group/generated-types.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
+// Generated file. DO NOT MODIFY IT BY HAND.
+
+export interface Payload {
+  /**
+   * Anonymous id
+   */
+  anonymousId: string
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/ripe/group/generated-types.ts:7]: hi @simpajj  anonymousId should be optional

File: packages/destination-actions/src/destinations/ripe/group/index.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+import { randomUUID } from 'crypto'
+
+const action: ActionDefinition<Settings, Payload> = {
+  title: 'Group',
+  description: 'Group user in Ripe',
+  defaultSubscription: 'type = "group"',
+  fields: {
+    anonymousId: {
+      type: 'string',
+      required: true,
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/ripe/group/index.ts:13]: @simpajj the anonymousId should be optional on a cloud mode Integration.

**Comment by joe-ayoub-segment** [comment]
hi @simpajj , 
Looks like the snapshot tests are still failing. If you can get them fixed, as well as the anonymousId being updated to optional then I can get the changes out tomorrow. 

image

**Comment by simpajj** [comment]
> hi @simpajj , Looks like the snapshot tests are still failing. If you can get them fixed, as well as the anonymousId being updated to optional then I can get the changes out tomorrow.
> 
> image

I've removed the snapshot tests for now, since they seem to generate new message ids and timestamps which end up always being different from the ones in the saved snapshot. Not sure how to get around that without not actually comparing the full snapshots, which then renders those tests a bit redundant.

**Comment by silesky** [comment]
> Not sure how to get around that without not actually comparing the full snapshots, which then renders those tests a bit redundant.

FYI, you can pass a property matcher as [the first argument to `toMatchSnapshot`](https://jestjs.io/docs/snapshot-testing#property-matchers)
```ts
it('will check the matchers and pass', () => {
  const user = {
    createdAt: new Date(),
    id: Math.floor(Math.random() * 20),
    name: 'LeBron James',
  };

  expect(user).toMatchSnapshot({
    createdAt: expect.any(Date),
    id: expect.any(Number),
  });
});

// Snapshot
exports[`will check the matchers and pass 1`] = `
Object {
  "createdAt": Any<Date>,
  "id": Any<Number>,
  "name": "LeBron James",
}
`;
```

**Comment by simpajj** [comment]
> @simpajj
> 
> > Not sure how to get around that without not actually comparing the full snapshots, which then renders those tests a bit redundant.
> 
> FYI, you can fix that mismatch issue by passing property matcher as [the first argument to `toMatchSnapshot`](https://jestjs.io/docs/snapshot-testing#property-matchers)
> 
> ```ts
> it('will check the matchers and pass', () => {
>   const user = {
>     createdAt: new Date(),
>     id: Math.floor(Math.random() * 20),
>     name: 'LeBron James',
>   };
> 
>   expect(user).toMatchSnapshot({
>     createdAt: expect.any(Date),
>     id: expect.any(Number),
>   });
> });
> 
> // Snapshot
> exports[`will check the matchers and pass 1`] = `
> Object {
>   "createdAt": Any<Date>,
>   "id": Any<Number>,
>   "name": "LeBron James",
> }
> `;
> ```

Thanks! Hopefully all good now :)

**Review by joe-ayoub-segment — APPROVED** [review]
Looks good. Tests pass, etc.

**Comment by joe-ayoub-segment** [comment]
hi @simpajj this code is now live. You can test, etc.

**Pull request merged**